106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,534 @@
+
+#### StrimziCruiseControlNotifier
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()`.
+The `StrimziCruiseControlNotifier` will implement the `AnomalyNotifier` interface.
+Every detected anomaly is placed in a priority queue and each of the anomalies have a priority value assigned to them. 
+Since we are working with only goal related violation every anomaly would have same priority, and they will be fixed on the basis of first come, first served mechanism.
+In case the anomaly is unfixable due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `StrimziCruiseControlNotifier` will override all the methods declared by the `AnomalyNotifier` interface.
+We will provide a concrete implementations of the `AnomalyNotifier.alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a ConfigMap with name set as `<cluster-name>-goal-violation-map`. 
+
+The ConfigMap will look like this:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: my-cluster-goal-violation-map
+data:
+  last_anomaly_detection_timestamp: <when last anomaly detection was run>
+  rebalance_finished_timestamp: <when the rebalance finished>
+  anomalyList: |
+    - <anomaly-id-1>
+    - <anomaly-id-2>
+    - <anomaly-id-3>    
+# ...
+```
+
+The advantages of using a single ConfigMap for every anomaly are:
+1. Single place to put in all the anomalies
+2. Even when dealing multiple Kafka clusters, the ConfigMap created will be equal to number of clusters deployed
+
+This generated config map will contain the list of goal violation anomalies in the section named `anomaly_list`. The ConfigMap will also store the timestamps for when the last anomaly detection was run and when the rebalance is finished.
+Whenever a rebalance is finished, the `anomalyList` would be set to empty by the operator. 
+
+The `last_anomaly_detection_timestamp` corresponds to when the last anomaly detection was run and will be updated everytime a new anomaly detection is run and the `rebalance_finished_timestamp` refers to the time when the running rebalance is finished and will be updated everytime when the running rebalance is finished.
+
+The created ConfigMap will persist when self-healing is enabled and will be deleted if the user decides to disable self-healing. This is because the ConfigMap is going to help us track the timestamps during the complete process.
+
+#### What happens if same anomaly is detected while the auto-rebalance is happening
```

Comment:
This section needs to be much clearer. Spell out the potential issue, the various scenarios, how logging and updating the time stamps will work and how that helps fix the issues under each scenario.